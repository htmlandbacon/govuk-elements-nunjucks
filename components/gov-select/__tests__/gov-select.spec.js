'use strict';

const nunjucks = require('nunjucks');
const cheerio = require('cheerio');

const options = [{value: 'one', display: 'One'}, {value: 'two', display: 'Two'}, {value: 'three', display: 'Three'}];

describe('gov-select', () => {
  it('should set the ID as select-{name} if ID attr not supplied and generate a select', () => {
    const name = 'selection';
    const output = nunjucks.render('./components/gov-select/template.njk', {
      label: 'select an option',
      options,
      name
    });
    const $ = cheerio.load(output);
    const selectID = $('select').attr('id');
    expect(selectID).toBe(`select-${name}`);
    expect(output).toMatchSnapshot();
  });

  it('should use ID attribute value over generated select-{name} ID and generate select with three options', () => {
    const id = 'my-id';
    const output = nunjucks.render('./components/gov-select/template.njk', {
      label: 'select an option',
      options,
      name: 'selection',
      id
    });
    const $ = cheerio.load(output);
    const selectID = $('select').attr('id');
    expect(selectID).toBe(`my-id`);

    const selectOptions = $('option');
    expect(selectOptions).toHaveLength(3);
    expect(selectOptions.get(0).attribs.value).toBe(`one`);
    expect(selectOptions.get(1).attribs.value).toBe(`two`);
    expect(output).toMatchSnapshot();
  });

  it('should add a form hint using the hint attribute', () => {
    const hint = 'Select one of the below';
    const output = nunjucks.render('./components/gov-select/template.njk', {
      label: 'select an option',
      options,
      name: 'selection',
      id: 'my-id',
      hint
    });
    const $ = cheerio.load(output);
    const hintText = $('label span.form-hint').text();
    expect(hintText).toBe(hint);

    const selectOptions = $('option');
    expect(selectOptions).toHaveLength(3);
    expect(selectOptions.get(0).attribs.value).toBe(`one`);
    expect(selectOptions.get(1).attribs.value).toBe(`two`);
    expect(output).toMatchSnapshot();
  });

  it('should set the option to selected if the value matches', () => {
    const value = 'two';
    const output = nunjucks.render('./components/gov-select/template.njk', {
      label: 'select an option',
      options,
      name: 'selection',
      id: 'my-id',
      value
    });
    const $ = cheerio.load(output);
    const selectOption = $('option');
    expect(selectOption).toHaveLength(3);
    const firstOption = selectOption.get(0).attribs;
    const secondOption = selectOption.get(1).attribs;
    expect(firstOption.value).toBe(`one`);
    expect(firstOption.selected).toBe(undefined);
    expect(secondOption.value).toBe(`two`);
    expect(secondOption.selected).toBe('');
    expect(output).toMatchSnapshot();
  });

  it('should add error message and classes when passed an error object', () => {
    const error = 'Select an option';
    const output = nunjucks.render('./components/gov-select/template.njk', {
      label: 'select an option',
      options,
      name: 'selection',
      id: 'my-id',
      error
    });
    const $ = cheerio.load(output);

    const formGroupClasses = $('.form-group').attr('class');
    const errorMsg = $('label span.error-message').text();

    expect(formGroupClasses).toBe('form-group form-group-error');
    expect(errorMsg).toBe(error);
    expect(output).toMatchSnapshot();
  });
  describe('select options', () => {
    it('should add mutiple options based on options and placeholder', () => {
      const value = 'three';
      const placeholder = 'placeholder';
      const output = nunjucks.render('./components/gov-select/template.njk', {
        label: 'select an option',
        options,
        placeholder,
        name: 'selection',
        id: 'my-id',
        value
      });
      const $ = cheerio.load(output);
      const selectOption = $('option');
      expect(selectOption).toHaveLength(4);
      const firstOption = selectOption.get(0).attribs;
      const secondOption = selectOption.get(1).attribs;
      const thirdOption = selectOption.get(2).attribs;
      const forthOption = selectOption.get(3).attribs;
      expect(firstOption.value).toBe(``);
      expect(firstOption.selected).toBe(undefined);
      expect(secondOption.value).toBe(`one`);
      expect(secondOption.selected).toBe(undefined);
      expect(thirdOption.value).toBe(`two`);
      expect(thirdOption.selected).toBe(undefined);
      expect(forthOption.value).toBe(`three`);
      expect(forthOption.selected).toBe('');
      expect(output).toMatchSnapshot();
    });
  });
  describe('class options', () => {
    it('should add classes to label span', () => {
      const classes = {label: 'test-label-class'};
      const output = nunjucks.render('./components/gov-select/template.njk', {
        label: 'select an option',
        options,
        name: 'selection',
        id: 'my-id',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('.form-group label span').attr('class');

      expect(labelClasses).toBe('test-label-class');
      expect(output).toMatchSnapshot();
    });
    it('should add classes to form-group div', () => {
      const classes = {fromGroup: 'test-group-class'};
      const output = nunjucks.render('./components/gov-select/template.njk', {
        label: 'select an option',
        options,
        name: 'selection',
        id: 'my-id',
        classes
      });
      const $ = cheerio.load(output);
      const formGroupClasses = $('.form-group').attr('class');
      expect(formGroupClasses).toBe('form-group test-group-class');
      expect(output).toMatchSnapshot();
    });
    it('should add classes to select span', () => {
      const classes = {selectClass: 'test-select-class'};
      const output = nunjucks.render('./components/gov-select/template.njk', {
        label: 'select an option',
        options,
        name: 'selection',
        id: 'my-id',
        classes
      });
      const $ = cheerio.load(output);
      const selectClasses = $('.form-control').attr('class');
      expect(selectClasses).toBe('form-control test-select-class');
      expect(output).toMatchSnapshot();
    });
  });
});
