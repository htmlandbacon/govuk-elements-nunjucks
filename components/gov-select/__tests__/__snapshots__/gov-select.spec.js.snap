// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gov-select class options should add classes to form-group div 1`] = `
<div id="my-id-form" class="form-group test-group-class">
  <label for="my-id">
    <span class="form-label-bold">select an option</span>
  </label>
  <select class="form-control" id="my-id" name="selection">
    <option value="one">One</option>
    <option value="two">Two</option>
    <option value="three">Three</option>
  </select>
</div>
`;

exports[`gov-select class options should add classes to label span 1`] = `
<div id="my-id-form" class="form-group">
  <label for="my-id">
    <span class="test-label-class">select an option</span>
  </label>
  <select class="form-control" id="my-id" name="selection">
    <option value="one">One</option>
    <option value="two">Two</option>
    <option value="three">Three</option>
  </select>
</div>
`;

exports[`gov-select class options should add classes to select span 1`] = `
<div id="my-id-form" class="form-group">
  <label for="my-id">
    <span class="form-label-bold">select an option</span>
  </label>
  <select class="form-control test-select-class" id="my-id" name="selection">
    <option value="one">One</option>
    <option value="two">Two</option>
    <option value="three">Three</option>
  </select>
</div>
`;

exports[`gov-select select options should add mutiple options based on options and placeholder 1`] = `
<div id="my-id-form" class="form-group">
  <label for="my-id">
    <span class="form-label-bold">select an option</span>
  </label>
  <select class="form-control" id="my-id" name="selection">
    <option value="">placeholder</option>
    <option value="one">One</option>
    <option value="two">Two</option>
    <option value="three" selected>Three</option>
  </select>
</div>
`;

exports[`gov-select should add a form hint using the hint attribute 1`] = `
<div id="my-id-form" class="form-group">
  <label for="my-id">
    <span class="form-label-bold">select an option</span>
    <span class="form-hint">Select one of the below</span>
  </label>
  <select class="form-control" id="my-id" name="selection">
    <option value="one">One</option>
    <option value="two">Two</option>
    <option value="three">Three</option>
  </select>
</div>
`;

exports[`gov-select should add error message and classes when passed an error object 1`] = `
<div id="my-id-form" class="form-group form-group-error">
  <label for="my-id">
    <span class="form-label-bold">select an option</span>
    <span class="error-message">Select an option</span>
  </label>
  <select class="form-control" id="my-id" name="selection">
    <option value="one">One</option>
    <option value="two">Two</option>
    <option value="three">Three</option>
  </select>
</div>
`;

exports[`gov-select should set the ID as select-{name} if ID attr not supplied and generate a select 1`] = `
<div id="select-selection-form" class="form-group">
  <label for="select-selection">
    <span class="form-label-bold">select an option</span>
  </label>
  <select class="form-control" id="select-selection" name="selection">
    <option value="one">One</option>
    <option value="two">Two</option>
    <option value="three">Three</option>
  </select>
</div>
`;

exports[`gov-select should set the option to selected if the value matches 1`] = `
<div id="my-id-form" class="form-group">
  <label for="my-id">
    <span class="form-label-bold">select an option</span>
  </label>
  <select class="form-control" id="my-id" name="selection">
    <option value="one">One</option>
    <option value="two" selected>Two</option>
    <option value="three">Three</option>
  </select>
</div>
`;

exports[`gov-select should use ID attribute value over generated select-{name} ID and generate select with three options 1`] = `
<div id="my-id-form" class="form-group">
  <label for="my-id">
    <span class="form-label-bold">select an option</span>
  </label>
  <select class="form-control" id="my-id" name="selection">
    <option value="one">One</option>
    <option value="two">Two</option>
    <option value="three">Three</option>
  </select>
</div>
`;
